### PR TITLE
Remove relevant metrics series on pod deletion (#23162).

### LINKED ISSUE #23162: CFP: Improvements in hubble metrics cardinality
## Cilium Feature Proposal/Improvement

Improvements in hubble metrics cardinality proposals:
1. introduce TTL on metrics
2. introduce limit on total number of metric series
3. clean up metrics on pod deletion
4. introduce label or annotation to exclude certain targets/pods from collecting metrics

**Is your feature request related to a problem?**

I'm currently examining potential hubble metrics cardinality issues, that may lead to OOM errors on either cilium-agent or metrics collector (in case of too many metrics are getting exposed by cilium agent).

Cilium-agent collects metrics in standard prometheus golang_client counters and depending on source & destination context labels metrics with pod-name (or cilium identity). Over time pods get deleted and new ones are created. Prometheus registry still keeps track of counters labeled with deleted pods. If cilium-agent doesn't get restarted, it can lead to OOM on cilium-agent or metrics collector. The problem is mostly visible when pods live for a very short time, eg. when heavy utilizing cronjobs or knative.

The other failing scenario is when a workload/pod behaves "strange". The port scanner may be an example.  Lets say we have a pod, that periodically scans open ports of other pods and clusters to report potential security vulnerabilities. Each scan attempt may generate thousands of dataseries with simple `nc -z -v <TARGET_IP> 1-32000`. It is related to the port-distribution metric https://github.com/cilium/cilium/blob/master/pkg/hubble/metrics/port-distribution/handler.go

**Describe the feature you'd like**

I can think of a few mitigations and improvements and would love to hear your feedback.

### 1. introduce TTL on metrics

Keep an additional map of the last write timestamp of each data series and remove metrics that haven't been updated in specified TTL. This will mitigate the problem of forever keeping dataseries of non-existing pods.
It can remove data series of existing posd when there is no frequent traffic.

### 2. introduce limit on total number of metric series

Introduce limit that will reject creating new metric series when the configured threshold is reached.  In combination with proposal (1) and (3) the number of data series may be later decreased which will allow creating new dataseries again.  The number of rejections should be tracked in a separate counter as a feedback to the maintainer.

This prevents OOM in any scenario, but may lead to some metrics not being collected.

### 3. clean up metrics on pod deletion

If metric is attributed to pod or CiliumIdentity we can track removals of these elements and remove related metrics. This solution is better than solution 1 in case of infrequent traffic.

On the other hand it can be quite challenging to determine which data series should be removed. eg. when source="pod-a" and destination="pod-b" should we remove such data series when "pod-a" is deleted? Should we wait for both pods to be deleted? 
If all pods talk to long-running pods like dns-server we may keep them forever (until dns pod is deleted).

### 4. introduce label or annotation to exclude certain targets/pods from collecting metrics

We can additionally label/annotate some pods with `cilium.io/no-metrics: true` label and in case of this annotation being present do not return any label in the context resolver. This will allow excluding certain workloads/pods from generating data series (pod scanner scenario).

### PR DESCRIPTION
**Pull request by marqc**
Signed-off-by: Marek Chodor <mchodor@google.com>

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [ ] Thanks for contributing!

Remove pod attributed hubble metrics after pod gets deleted.

Fixes: #23162

### CODE REVIEW

File: pkg/hubble/metrics/api/api.go
Comment (christarazi) [inline pkg/hubble/metrics/api/api.go]: Nit: No need for extra line here

File: pkg/hubble/metrics/api/context.go
Comment (christarazi) [inline pkg/hubble/metrics/api/context.go]: We probably want to preallocate the `sourceCtxCfg` and `destinationCtxCfg`, 

```suggestion
sourceCtxCfg := make([]type, 0, len(o.SourceEgress) + len(o.SourceIngress))
```

File: pkg/hubble/metrics/metrics.go
Comment (christarazi) [inline pkg/hubble/metrics/metrics.go]: I think we should avoid using global variables at the package level like this. Consts are other insignificant variables are fine at the package level. Would you mind refactoring this to be be part of some struct that we instantiate, etc.?

**Review by christarazi — CHANGES_REQUESTED** [review]
Thanks for the PR. A few comments below.

File: pkg/hubble/metrics/api/api.go
Comment (chancez) [inline pkg/hubble/metrics/api/api.go]: I believe this is the wrong design. Instead we should have an interface to get a list of metricsVec's from a handler. Then we can loop over every handler, and every MetricVector for each handler, calling `DeleteMetricsAssociatedWithPod`, instead of making every handler handle this explicitly.

File: pkg/hubble/metrics/api/context.go
Comment (chancez) [inline pkg/hubble/metrics/api/context.go:541]: This should probably handle `ContextPodShort` and `ContextLabels`. For `ContextLabels` it should check if `source_pod`, or `destination_pod` values were configured, and if so, check their values.

File: pkg/hubble/metrics/metrics.go
Comment (chancez) [inline pkg/hubble/metrics/metrics.go:84]: There is nothing ensuring this go routine gets cleaned up, we should be waiting on it somewhere.

File: pkg/hubble/metrics/metrics.go
Comment (chancez) [inline pkg/hubble/metrics/metrics.go]: This blocks deletions of all other handler metrics series. Instead we could probably use a proper queue type which lets us order things by when we want to delete them, rather than a channel. The `k8s.io/client-go/util/workqueue` is a good package for doing this that we already use in cilium.

File: pkg/hubble/metrics/metrics.go
Comment (chancez) [inline pkg/hubble/metrics/metrics.go:52]: Similar to @christarazi 's comment, I wonder if this could be a method on some object instead, and then the queue would be a field on a struct rather than a global.

File: pkg/hubble/metrics/api/context.go
Comment (marqc) [reply pkg/hubble/metrics/api/context.go:541]: +1 for ContextLabels, but I believe it should only remove data series when both "source_pod+source_namespace" or "destination_pod+destination_namespace" is set.

I disagree with handling "ContextPodShort". There is a possibility that multiple pods from single deployment runs on a single k8s node. If we remove dataseries on pod deletion configured with ContextPodShort, we would remove it when one of the pods gets deleted and others are not affected. This will be definitely  counter-intuitive and not expected for users.

File: pkg/hubble/metrics/metrics.go
Comment (marqc) [reply pkg/hubble/metrics/metrics.go]: done

File: pkg/hubble/metrics/metrics.go
Comment (marqc) [reply pkg/hubble/metrics/metrics.go]: Refactored a little bit, but there is still a place for improvements (eg. setting delay in config options)

File: pkg/hubble/metrics/api/context.go
Comment (marqc) [reply pkg/hubble/metrics/api/context.go]: refactored to calculate it once on context initialization instead of every PodDeletion handling

File: pkg/hubble/metrics/api/api.go
Comment (marqc) [reply pkg/hubble/metrics/api/api.go]: Refactored in a way you suggested.

File: pkg/hubble/metrics/metrics.go
Comment (marqc) [reply pkg/hubble/metrics/metrics.go:84]: I changed the implementation to workqueue and right now goroutine will exit after calling `queue.Shutdown()` but I'm not sure where to put it. A similar goroutine for webserver with metrics seems to not have anything that could stop it. Can you point me where to implement this "shutdown hook"?

File: pkg/hubble/metrics/api/context.go
Comment (chancez) [reply pkg/hubble/metrics/api/context.go:541]: Your right, PodShort is basically "WorkloadName", but not as reliable.

File: pkg/hubble/metrics/metrics.go
Comment (chancez) [reply pkg/hubble/metrics/metrics.go:84]: Good point, I didn't notice that. I guess we don't do properly handle that either. It would require refactoring most of the code in this package to work on a struct with methods. We can probably we can fix this later, I think it's a bit unrelated to this PR.

**Review by chancez — APPROVED** [review]
This looks great, I think I'd like to have another hubble maintainer review it also though.

**Review by christarazi — APPROVED** [review]

**Review by sayboras — APPROVED** [review]
Thanks and LGTM :heavy_check_mark:

File: pkg/hubble/metrics/api/api.go
Comment (gandro) [inline pkg/hubble/metrics/api/api.go]: Should we do a `nil` check on `Context()` here? Otherwise we should document that `Context()` must not return nil.

**Review by gandro — APPROVED** [review]
I'm not super familiar with the Prometheus SDK, but the Hubble API changes look good to me.

File: pkg/hubble/metrics/api/api.go
Comment (marqc) [reply pkg/hubble/metrics/api/api.go]: fixed

File: pkg/hubble/metrics/api/registry_test.go
Comment (ghost) [inline pkg/hubble/metrics/api/registry_test.go]: Not critical, but would be nice to have similar tests for different set of options, e.g. `sourceContext=workload-name;destinationContext=workload-name;labelsContext=source_pod,source_namespace,destination_pod,destination_namespace` (metrics will be deleted) or `sourceContext=identity;destinationContext=identity;labelsContext=source_pod,destination_pod` (metrics won't be deleted, even though they can be associated with pods).

**Review by ghost — COMMENTED** [review]
I think it should be documented that metrics are cleaned up on pod deletion and for which options. It's an internal detail, but an important one, and it's not entirely obvious options you need to set to get this behaviour.

File: pkg/hubble/metrics/api/registry_test.go
Comment (marqc) [reply pkg/hubble/metrics/api/registry_test.go]: Good point! New test cases added.

**Comment by marqc** [comment]
> I think it should be documented that metrics are cleaned up on pod deletion and for which options. It's an internal detail, but an important one, and it's not entirely obvious options you need to set to get this behaviour.

I added a note in metrics Documentation.

File: Documentation/observability/metrics.rst
Comment (ghost) [inline Documentation/observability/metrics.rst]:
```suggestion
    * ``labelsContext`` contains both ``source_namespace`` and ``source_pod`` and metric series labels match namespace and name of deleted pod
```

File: Documentation/observability/metrics.rst
Comment (ghost) [inline Documentation/observability/metrics.rst]:
```suggestion
    * ``labelsContext`` contains both ``destination_namespace`` and ``destination_pod`` and metric series labels match namespace and name of deleted pod
```

**Review by ghost — APPROVED** [review]
Docs nit, but looks good to me, thanks!

**Review by qmonnet — CHANGES_REQUESTED** [review]
The changes look good, thanks!

Please, do not remove review assignments. (I'm not re-adding Aditi because you already have reviews from members of sig-policy.)

image

This is not the case. The change in your PR is not trivial, so I would really appreciate some more detailed context and motivation in the commit description. It should be possible to understand the issue and how it's addressed just by looking at the Git logs, without having to look up for the GitHub issue. Nit: The issue fixed by the commit should be in a `Fixes:` tag at the end of the description, rather than in the commit object.

**Comment by marqc** [comment]
> Please, do not remove review assignments. (I'm not re-adding Aditi because you already have reviews from members of sig-policy.)

Sorry. It looks like github automatically removes assignments when I click on re-request review.

> This is not the case. The change in your PR is not trivial, so I would really appreciate some more detailed context and motivation in the commit description. It should be possible to understand the issue and how it's addressed just by looking at the Git logs, without having to look up for the GitHub issue. Nit: The issue fixed by the commit should be in a `Fixes:` tag at the end of the description, rather than in the commit object.

I updated the commit message.

**Review by qmonnet — APPROVED** [review]
> Sorry. It looks like github automatically removes assignments when I click on re-request review.

OK, I didn't know that

> I updated the commit message.

Great description, thanks a lot!

**Comment by qmonnet** [comment]
Reviews are in and the tests passed eventually! Deferring to TopHat for the rest of the operations. Thanks!

**Comment by pchaigno** [comment]
Looks like Travis CI failed to run. Do we know why? If not, we may need to rebase and rerun CI :disappointed:

**Comment by qmonnet** [comment]
> Looks like Travis CI failed to run. Do we know why?

I don't :/. Scrolling back until 26th of January [on Travis](https://app.travis-ci.com/github/cilium/cilium/pull_requests) (couldn't find a way to filter based on PRs), I can't find any build for this PR. Not sure what happened.

**Comment by pchaigno** [comment]
I tried to close and reopen to trigger the Travis CI job but that didn't seem to work.
@marqc Could you rebase your branch to see if it helps? Sorry for the trouble :confused:

**Comment by marqc** [comment]
> I tried to close and reopen to trigger the Travis CI job but that didn't seem to work. @marqc Could you rebase your branch to see if it helps? Sorry for the trouble 😕

@pchaigno done

**Comment by qmonnet** [comment]
/test-1.16-4.19

**Comment by pchaigno** [comment]
At this point, I suspect it's something weird with this branch, @marqc's fork, or @marqc's account itself :shrug: I've opened https://github.com/cilium/cilium/pull/23900 with these changes, just to run the Travis CI job.

**Comment by qmonnet** [comment]
Here it is at last https://github.com/cilium/cilium/pull/23385/checks?check_run_id=11488636578

**Comment by qmonnet** [comment]
/test-1.16-4.19

**Comment by pchaigno** [comment]
/test-1.16-4.19

**Pull request merged**
